filter/series_by_tag.go

```diff
@@ -114,42 +116,95 @@ func createMatcher(spec TagSpec) func(string, map[string]string) bool {
 	}
 }
 
+type MatchingFunc func(string, map[string]string) bool
+
 // SeriesByTagPatternIndex helps to index the seriesByTag patterns and allows to match them by metric
 type SeriesByTagPatternIndex struct {
-	patternToMatcher map[string]func(string, map[string]string) bool
+	namesPatternIndex      *PatternIndex
+	nameToPatternMatchers  map[string]map[string]MatchingFunc
```

Comment:
I think it would be better to move `MatchingFunc` to `PatternNode` structure and remove `nameToPatternMatchers`.

Reply:
I will do it in the next PR.